torchopt/alias/adagrad.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2022 MetaOPT Team. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+# This file is modified from:
+# https://github.com/deepmind/optax/blob/master/optax/_src/alias.py
+# ==============================================================================
+# Copyright 2019 DeepMind Technologies Limited. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Preset :class:`GradientTransformation` for the AdaGrad optimizer."""
+
+from torchopt.alias.utils import flip_sign_and_add_weight_decay, scale_by_neg_lr
+from torchopt.combine import chain
+from torchopt.schedule import polynomial_schedule
+from torchopt.transform import scale_by_rss
+from torchopt.typing import GradientTransformation, Scalar
+
+
+__all__ = ['adagrad']
+
+
+# pylint: disable-next=too-many-arguments
+def adagrad(
+    lr: Scalar = 1e-2,
+    lr_decay: float = 0.0,
+    weight_decay: float = 0.0,
+    initial_accumulator_value: float = 0.0,
+    eps: float = 1e-10,
+    *,
+    maximize: bool = False,
+) -> GradientTransformation:
+    """The functional AdaGrad optimizer.
+
+    AdaGrad is an algorithm for gradient based optimization that anneals the learning rate for each
+    parameter during the course of training.
+    WARNING: AdaGrad's main limit is the monotonic accumulation of squared gradients in the
+    denominator: since all terms are >0, the sum keeps growing during training and the learning rate
+    eventually becomes very small.
```

Comment:
```suggestion

    .. warning::
        AdaGrad's main limit is the monotonic accumulation of squared gradients in the denominator.
        Since all terms are ``> 0``, the sum keeps growing during training, and the learning rate
        eventually becomes very small.
```